src/Microsoft.AspNetCore.Mvc.Razor/Internal/FileProviderRazorProject.cs

```diff
@@ -13,22 +14,37 @@ public class FileProviderRazorProject : RazorProject
     {
         private const string RazorFileExtension = ".cshtml";
         private readonly IFileProvider _provider;
+        private readonly IHostingEnvironment _hostingEnvironment;
 
-        public FileProviderRazorProject(IRazorViewEngineFileProviderAccessor accessor)
+        public FileProviderRazorProject(IRazorViewEngineFileProviderAccessor accessor, IHostingEnvironment hostingEnviroment)
         {
             if (accessor == null)
             {
                 throw new ArgumentNullException(nameof(accessor));
             }
 
+            if (hostingEnviroment == null)
+            {
+                throw new ArgumentNullException(nameof(hostingEnviroment));
+            }
+
             _provider = accessor.FileProvider;
+            _hostingEnvironment = hostingEnviroment;
         }
 
         public override RazorProjectItem GetItem(string path)
         {
             path = NormalizeAndEnsureValidPath(path);
             var fileInfo = _provider.GetFileInfo(path);
-            return new FileProviderRazorProjectItem(fileInfo, basePath: string.Empty, path: path);
+
+            string relativePhysicalPath = null;
+            if (fileInfo != null && fileInfo.Exists)
+            {
+                var absoluteBasePath = _hostingEnvironment.ContentRootPath;
+                relativePhysicalPath = fileInfo?.PhysicalPath?.Substring(absoluteBasePath.Length + 1); // Include leading separator
```

Comment:
We'll need to ensure the path actually starts with `ContentRootPath`

Reply:
Also, could we calculate this lazily? Presumably we don't use the relativePath in the common case where things just work?